src/transformers/modeling_utils.py

```diff
@@ -679,36 +679,13 @@ def _load_state_dict_into_model(model_to_load, state_dict, start_prefix):
 
     error_msgs = []
 
-    # PyTorch's `_load_from_state_dict` does not copy parameters in a module's descendants
-    # so we need to apply the function recursively.
-    def load(module: nn.Module, state_dict, prefix=""):
-        local_metadata = {} if metadata is None else metadata.get(prefix[:-1], {})
-        args = (state_dict, prefix, local_metadata, True, [], [], error_msgs)
-        # Parameters of module and children will start with prefix. We can exit early if there are none in this
-        # state_dict
-        if len([key for key in state_dict if key.startswith(prefix)]) > 0:
-            if is_deepspeed_zero3_enabled():
-                import deepspeed
-
-                # In sharded models, each shard has only part of the full state_dict, so only gather
-                # parameters that are in the current state_dict.
-                named_parameters = dict(module.named_parameters(prefix=prefix[:-1], recurse=False))
-                params_to_gather = [named_parameters[k] for k in state_dict.keys() if k in named_parameters]
-                if len(params_to_gather) > 0:
-                    # because zero3 puts placeholders in model params, this context
-                    # manager gathers (unpartitions) the params of the current layer, then loads from
-                    # the state dict and then re-partitions them again
-                    with deepspeed.zero.GatheredParameters(params_to_gather, modifier_rank=0):
-                        if torch.distributed.get_rank() == 0:
-                            module._load_from_state_dict(*args)
```

Comment:
Have we tested the new code with deepspeed too?

Reply:
Not yet! To come next week :)

Reply:
More than likely though I just leave this here to enable and then we move on, it was interim me cutting it *all* out

Reply:
I don't think this can work if the model has already been sharded under deepspeed, because deepspeed hijacks the tensors and the loading will fail.

But I'd suggest to check in with the deepspeed team - perhaps they have some more recent tricks that will accomplish that faster.

Reply:
But in general - under `zero.Init` the model is already spread out across the gpus, so you can't just overwrite its shards - w/o the machinery you propose to delete.

I think the just published Universal Checkpoint might be usable here to broadcast the updated tensor shards to each gpu, w/o needing to gather their content first.

@tjruwase, if possible could you please assist if there is a way to update the already sharded tensors in a faster way?

Reply:
of course if this doesn't work, then you'd need to have 2 code branches. The users who use `zero.Init` won't mind waiting, because the huge models they want to load won't load onto a single gpu, so cost of slower loading speed is a trade off here.

Reply:
> @tjruwase, if possible could you please assist if there is a way to update the already sharded tensors in a faster way?

@stas00, I am just catching up on this, but my initial thought is wouldn't the following feature help for this case?
https://deepspeed.readthedocs.io/en/latest/zero3.html#modifying-partitioned-states